### PR TITLE
Add font family and text-decoration typography supports to paragraph blocks

### PR DESCRIPTION
**Pull request by randhirexpresstech**
This PR adds some missing typography like font family and text-decoration supports to paragraph block, which are already available in Gutenberg. 
With the help of this PR block theme developers and website admin will have more customization options to create more beautiful website and design patterns. 
In some places, designers require to have different typography as compare to global typography like font-family, text-decoration etc. to create unique designs. 
In the absence of these typography settings on block level, block theme developers are rely on custom css for these typography options which are not editable by website admin using frontend editor tools. So this PR will provide more flexibility and help admin to get more controls over design and UI elements.

Testing Steps
1. Activate Twenty Twenty-Two theme
2. Add a new page
3. Add a paragraph. 
4. Assign font-family and text-decoration.
5. Publish the page 
6. View and check the page for assigned typography options 

Please check this [screen-cast](https://app.usebubbles.com/tY4CcWhrrUzWa6KafMmP7A/typography-supports-to-button-heading-and-paragraph-blocks)
For more details please check this [comment](https://github.com/WordPress/gutenberg/pull/39642#pullrequestreview-1079909583)

### CODE REVIEW

**Comment by ramonjd** [comment]
Related discussions on typography defaults over at https://github.com/WordPress/gutenberg/issues/35604

**Comment by talldan** [comment]
Thanks for working on this @randhirexpresstech. I've added some reviewers who will be more aware of which blocks are ready for which particular features.

File: packages/block-library/src/button/block.json
Comment (amustaque97) [inline packages/block-library/src/button/block.json]: Letter spacing should not allow negative values. Here is an example how it looks if letter space is a negative value.
image

**Comment by amustaque97** [comment]
Hi @randhirexpresstech, thank you for working on this. 🥳 There is a minor review comment from my side. Rest of the changes looks good to me. 😃

**Comment by amustaque97** [comment]
Also I'm not 100% sure if this is out of scope of the current PR or not. It would be really appreciated if we can show what **default font** is? I'm sharing few screenshots both for paragraph and heading.I got confused why style didn't change when I changed from _default font to the same actual font_. Another benefit is user will not waste time by selecting same default font again. 😅  

**Heading**
Screenshot 2022-04-02 at 9 48 33 AM

**Paragraph**
Screenshot 2022-04-02 at 9 52 14 AM

cc @ntsekouras @paaljoachim

**Comment by ntsekouras** [comment]
Thanks for your work here @randhirexpresstech! There is some prior art to this like [in this PR](https://github.com/WordPress/gutenberg/pull/37586), which added font family support in `paragraph` and then reverted this change.

The main reasons might be these ([comment](https://github.com/WordPress/gutenberg/pull/34064#issuecomment-959384115)): 

>1. We should keep font family as an option if it exists, but it should not be the default for any blocks
>2. We don't yet have a good way to add custom fonts, so it's quite limited.
>3. Once we have a better fonts API we can surface this more in blocks that are more likely to need font family changes (site title being an obvious one).

Regarding the `button` block there are folks that worked/discussed this more.. --cc @apeatling

File: packages/block-library/src/button/block.json
Comment (ntsekouras) [reply packages/block-library/src/button/block.json]: I think this needs a new issue(if one not exists). Could you check and create one @amustaque97 ?

Comment (amustaque97) [reply packages/block-library/src/button/block.json]: Created an issue https://github.com/WordPress/gutenberg/issues/39993

Comment (amustaque97) [reply packages/block-library/src/button/block.json]: Closed the issue #39993 and this review comment is no longer valid. 🙂

**Comment by apeatling** [comment]
@mtias Since we have the fonts API in place now, can we continue adding these supports, assuming they don't appear by default?

**Comment by mtias** [comment]
There's a few details of the integration between registered fonts and theme.json enqueued fonts that needs to be sorted before showing font-family more prominently in block instances.

**Comment by luminuu** [comment]
@mtias Now that the font family is available for headings, can it also be made available for paragraphs now? Or are there still issues around this?

**Comment by ndiego** [comment]
Assuming there are not technical blockers, and given that `font-family` is now supported on Heading I assume there isn't, now would be a great time to get this added to Paragraphs. It's one of the most common issues I hear in the community. Once implemented, we can continue implementation across other text-based blocks.

**Comment by ndiego** [comment]
@randhirexpresstech since `font-family` has already been added to Gutenberg in #40970, would you be willing to update this PR to only focus on adding `font-family` to Paragraphs? If approved to merge, I want to make sure you get credit for your contribution. Thanks!

**Comment by aaronrobertshaw** [comment]
Not to rain on anyone's parade but I'm not sure the issues around the fonts API @mtias mentioned have been fully resolved yet. Having a look around there are still a [number of open issues](https://github.com/WordPress/gutenberg/labels/%5BFeature%5D%20Webfonts) (assuming I'm looking in the right places 😄 ).

> There's a few details of the integration between registered fonts and theme.json enqueued fonts that needs to be sorted before showing font-family more prominently in block instances.

Perhaps, Matías can confirm we're safe to proceed with adding these supports now?

> Assuming there are not technical blockers, and given that font-family is now supported on Heading I assume there isn't

I don't think the addition of the font-family support to the Heading block considered the current pause on adding this block support and might have been slightly rushed through. That said, an argument could be made for that block the lack of consistency with the other title blocks was a greater issue.

I know this is an easy change and a lot of people (myself included) wish to see block supports like font-family being leveraged more widely but we might only need to be patient for a little while longer 🤞

**Comment by randhirexpresstech** [comment]
> @randhirexpresstech since `font-family` has already been added to Gutenberg in #40970, would you be willing to update this PR to only focus on adding `font-family` to Paragraphs? If approved to merge, I want to make sure you get credit for your contribution. Thanks!

Thanks @ndiego  for having a look into this PR. I think typography supports (e.g. Line height, Letter spacing, Text transform, Font style, Font weight ) should be consistent among all text based blocks (button, heading and paragraph are too common to use). I have raised this PR because i have gone through this issue while i was doing some design work using these text based block and thought it may be useful for others. Having said that, I am happy to change the PR to just support font-family for paragraph if your idea to is to keep this simple and raise [another PR for typography supports to button block](https://github.com/WordPress/gutenberg/pull/41435).

**Comment by marctison75** [comment]
@mtias would you please confirm that we can use the new font-family setting for the Heading block #40970, and that it won't be reverted in a future Gutenberg release?

**Comment by ndiego** [comment]
Thanks @randhirexpresstech, I updated the title accordingly. Generally, it is easier to break out enhancements per block. 

Can someone point me to the technical issues that are blocking the addition of `font-family` to typography controls? We already have `font-family` on a number of Core blocks, not including the Heading block. I am not aware of any issues impacting those blocks that support `font-family` but obviously could be missing something. If there are issues, I would like to help get them fixed. This would be a HUGE win for the community to expand typography controls.

**Comment by andrewserong** [comment]
@ndiego, if I'm reading the earlier comments right, the issue to opting-in more blocks to the `font-family` support was due to the Webfonts API's enqueuing support (this comment: https://github.com/WordPress/gutenberg/pull/39642#issuecomment-1112525791)

It looks like the Webfonts API work is being tracked in this issue: https://github.com/WordPress/gutenberg/issues/41479 and the programmatic enqueueing is in https://github.com/WordPress/gutenberg/issues/40363 and https://github.com/WordPress/gutenberg/issues/40386 — if I understand correctly from those issue descriptions, the problem with opting-in prematurely could be that for a user who has access to the control to select the font, when they interact with the control, the font they've selected may not actually be available / enqueued in the editor at that time, so they will not see the appropriate font that they've selected.

I haven't been involved in working in that particular area, but just thought I'd share those links in case it helps give a bit more context to the discussions (I've only done a quick read of the issues, I haven't tested out the current state of things yet 🙂)

**Comment by ndiego** [comment]
@mtias and/or @hellofromtonya would either of you be able to provide any clarification on the discussion above. I assume that for 6.1 we want to have fonts stabilized and at a place where we can enable `font-family` across nearly all blocks. This has been a very common "ask" in the community. 

Currently, we have `font-family` enabled across a handful of blocks but obviously not all. Time is slipping by quickly and I would like to have PRs ready to go for all blocks we want to have `font-family` support, so as soon as the necessary font API PRs are merged, we can have block supports updated accordingly. Let me know what you think!

**Comment by ndiego** [comment]
Just created this tracking issue for `font-family` support. It's interesting to see how many blocks do and don't support this feature 🤔 https://github.com/WordPress/gutenberg/issues/42646

**Comment by aaronrobertshaw** [comment]
🎉  We've got a green light to add missing typography supports as part of greater efforts to achieve consistency in our design tools. 

If it interests you, you can track the overall progress towards consistent block support adoption in: https://github.com/WordPress/gutenberg/issues/43241. Or, there's a [tracking issue](https://github.com/WordPress/gutenberg/issues/43242) specifically for typography supports.

With that said, @randhirexpresstech, would you mind updating this PR also to include adding the missing text-decoration support to the paragraph block? i.e. `"__experimentalTextDecoration": true,`

If you are short on time, I'd be happy to push the required changes or create an alternate PR. Thanks!

**Comment by mtias** [comment]
The issues with the webfonts API are related to the distinction of setting a font family for a block type and setting it for a block instance. Block instances should only get access to already vetted fonts by the global styles declaration, and users should not be encouraged to setting font-family locally, which is why it should not be visible by default (ellipsis menu) on block instances. For block types, access to font family should be consistent and clear, and we also need to expand the reporting of what blocks are using what font on the global styles typography settings.

**Comment by randhirexpresstech** [comment]
> tada We've got a green light to add missing typography supports as part of greater efforts to achieve consistency in our design tools.
> 
> If it interests you, you can track the overall progress towards consistent block support adoption in: #43241. Or, there's a [tracking issue](https://github.com/WordPress/gutenberg/issues/43242) specifically for typography supports.
> 
> With that said, @randhirexpresstech, would you mind updating this PR also to include adding the missing text-decoration support to the paragraph block? i.e. `"__experimentalTextDecoration": true,`
> 
> If you are short on time, I'd be happy to push the required changes or create an alternate PR. Thanks!

Thanks @aaronrobertshaw for the good news and providing me such a great opportunity to contribute in WordPress [gutenberg](https://github.com/WordPress/gutenberg) repository. I am happy to update this PR to include adding the missing text-decoration support to the paragraph block? i.e. "__experimentalTextDecoration": true, as per requirement.

**Review by aaronrobertshaw — APPROVED** [review]
Thanks for updating this PR @randhirexpresstech 👍 

This is testing well for me.

✅ Font family and text decoration selections are applied in the block editor
✅ Selections are reflected on the frontend
✅ New controls are not displayed by default (they're only available via the ellipsis menu)
✅ Font family is available by default in the paragraph block's Global Styles panel

<details>
<summary>Click for block editor demo</summary>

https://user-images.githubusercontent.com/60436221/185850514-645d584a-6135-4485-a349-60cb73e0b4a7.mp4
</details>



<details>
<summary>Click for Paragraph block's Global Styles panel demo</summary>


https://user-images.githubusercontent.com/60436221/185850483-1c498250-56ce-435f-a4d0-eec2d37914a6.mp4
</details>

The failed e2e tests are currently being re-run. When they pass this should almost be ready to land. 

@randhirexpresstech It might also help for clarity and posterity if you could update the PR description here to reflect the actual changes in the current version of the PR i.e. make the description about updating the paragraph typography block supports.

**Comment by randhirexpresstech** [comment]
> Thanks for updating this PR @randhirexpresstech +1
> 
> This is testing well for me.
> 
> white_check_mark Font family and text decoration selections are applied in the block editor white_check_mark Selections are reflected on the frontend white_check_mark New controls are not displayed by default (they're only available via the ellipsis menu) white_check_mark Font family is available by default in the paragraph block's Global Styles panel
> 
> Click for block editor demo
> Click for Paragraph block's Global Styles panel demo
> The failed e2e tests are currently being re-run. When they pass this should almost be ready to land.
> 
> @randhirexpresstech It might also help for clarity and posterity if you could update the PR description here to reflect the actual changes in the current version of the PR i.e. make the description about updating the paragraph typography block supports.

Thanks @aaronrobertshaw for approving this PR. As per requirement, I have updated title and description of this PR for more clarity.

**Pull request merged**
